openvasd/openvasd.c

```diff
@@ -0,0 +1,2297 @@
+ * metadata chunk by chunk.
+ *
+ * @param conn Connector struct with the data necessary for the connection
+ * @param mhnd The curl multiperform handler
+ * @param resp The stringstream struct for the write callback function.
+ *
+ * @return The response. Null on error.
+ */
+openvasd_resp_t
+openvasd_get_vts_stream_init (openvasd_connector_t *conn,
+                              openvasd_curlm_t *mhnd,
+                              openvasd_stringstream *resp)
+{
+  GString *path;
+  openvasd_resp_t response = NULL;
+  gchar *err = NULL;
+  CURL *hnd = NULL;
+  CURLM *h = NULL;
+  struct curl_slist *customheader = NULL;
+
+  response = g_malloc0 (sizeof (struct openvasd_response));
+
+  path = g_string_new ("/vts?information=1");
+  customheader = init_customheader ((*conn)->apikey, FALSE);
+  if ((hnd = handler (conn, GET, path->str, NULL, customheader, resp, &err))
+      == NULL)
+    {
+      g_string_free (path, TRUE);
+      g_free (resp->ptr);
+      response->code = RESP_CODE_ERR;
+      response->body = err;
+      return response;
+    }
+  g_string_free (path, TRUE);
+
+  h = curl_multi_init ();
+  curl_multi_add_handle (h, hnd);
+
+  if (mhnd == NULL)
+    mhnd = openvasd_curlm_handler_new ();
+
+  mhnd->h = h;
+  mhnd->customheader = customheader;
+
+  response->code = RESP_CODE_OK;
+  return response;
+}
+
+/**
+ * @brief Get a new feed metadata chunk.
+ *
+ * This function must be call until the
+ * return value is 0, meaning there is no more data to fetch.
+ *
+ * @param mhnd Curl multiperfom for requesting the feed metadata
+ *
+ * @return greather than 0 if the handler is still getting data. 0 if the
+ * transmision finished. -1 on error
+ */
+int
+openvasd_get_vts_stream (openvasd_curlm_t *mhnd)
+{
+  static int running = 0;
+  CURLM *h = mhnd->h;
+  if (!(h))
+    {
+      return -1;
+    }
+
+  CURLMcode mc = curl_multi_perform (h, &running);
+  if (!mc && running)
+    /* wait for activity, timeout or "nothing" */
+    mc = curl_multi_poll (h, NULL, 0, 5000, NULL);
+  if (mc != CURLM_OK)
+    {
+      g_warning ("%s: error on curl_multi_poll(): %d\n", __func__, mc);
+      return -1;
+    }
+
+  return running;
+}
+
+/**
+ * @brief Get VT's metadata
+ *
+ * @param conn Connector struct with the data necessary for the connection
+ *
+ * @return Response Struct containing the feed metadata in json format in the
+ * body. NULL on error.
+ */
+openvasd_resp_t
+openvasd_get_vts (openvasd_connector_t *conn)
+{
+  GString *path;
+  openvasd_resp_t response = NULL;
+  gchar *err = NULL;
+  CURL *hnd = NULL;
+  openvasd_stringstream resp;
+  struct curl_slist *customheader = NULL;
+
+  response = g_malloc0 (sizeof (struct openvasd_response));
+
+  init_openvasd_stringstream (&resp);
+  path = g_string_new ("/vts?information=1");
+  customheader = init_customheader ((*conn)->apikey, FALSE);
+  if ((hnd = handler (conn, GET, path->str, NULL, customheader, &resp, &err))
+      == NULL)
+    {
```

Comment:
needs to free customheader

Reply:
same for all following cases

Reply:
Maybe put the customheader in the struct that is currently called openvas_stringstream. Then init them together, and free them together with a dedicated free function. Would also be neater than calling `g_free (resp.ptr)` to free something created with `init_openvas_stringstream`.